common/cache/metricsScopeCache.go

```diff
@@ -23,50 +23,122 @@
 package cache
 
 import (
-	"bytes"
-	"strconv"
 	"sync"
+	"sync/atomic"
+	"time"
 
+	"github.com/uber/cadence/common"
 	"github.com/uber/cadence/common/metrics"
 )
 
-type domainMetricsScopeCache struct {
-	sync.RWMutex
-	scopeMap map[string]metrics.Scope
-}
+const flushBufferedMetricsScopeDuration = 10 * time.Second
+
+type (
+	metricsScopeMap map[string]map[int]metrics.Scope
+
+	buffer struct {
+		sync.RWMutex
+		bufferMap metricsScopeMap
+	}
+
+	domainMetricsScopeCache struct {
+		status        int32
+		buffer        *buffer
+		cache         atomic.Value
+		closeCh       chan struct{}
+		flushDuration time.Duration
+	}
+)
 
 // NewDomainMetricsScopeCache constructs a new domainMetricsScopeCache
 func NewDomainMetricsScopeCache() DomainMetricsScopeCache {
-	return &domainMetricsScopeCache{
-		scopeMap: make(map[string]metrics.Scope),
+
+	mc := &domainMetricsScopeCache{
+		buffer: &buffer{
+			bufferMap: make(metricsScopeMap),
+		},
+		closeCh:       make(chan struct{}),
+		flushDuration: flushBufferedMetricsScopeDuration,
+	}
+
+	mc.cache.Store(make(metricsScopeMap))
+	return mc
+}
+
+func (c *domainMetricsScopeCache) flushBufferedMetricsScope(flushDuration time.Duration) {
+	for {
+		select {
+		case <-time.After(flushDuration):
+			c.buffer.Lock()
+			if len(c.buffer.bufferMap) > 0 {
+				scopeMap := make(metricsScopeMap)
+
+				data := c.cache.Load().(metricsScopeMap)
+				// Copy everything over after atomic load
+				for key, val := range data {
+					scopeMap[key] = val
+				}
+
+				// Copy from buffered array
+				for key, val := range c.buffer.bufferMap {
+					if _, ok := scopeMap[key]; !ok {
+						scopeMap[key] = map[int]metrics.Scope{}
+					}
+					for k, v := range val {
+						scopeMap[key][k] = v
```

Comment:
Will this result in race condition? I mean the value got copied on L79 is a map reference, this map will be read by the `Get()` method and be updated here concurrently I think? My understanding is that in this case whatever value got stored to the `atomic.Value` should be read only.

I am not sure why the concurrent test doesn't report data race though.

Reply:
oh, i see, then copy everything over will fix this.

Reply:
Updated.